TLCALIPER-124 send caliper 1.1 events to Endpoint

pushyamig opened the pull request:
This Effort is tracked by the https://itsjira.umms.med.umich.edu/browse/TLCALIPER-124
This PR is reading the caliper 1.1 events from the directory given from the command line modifying the caliper event timestamp and sending to an Endpoint. This script is tested to send events to the http://lti.tools/caliper/event?key=udp_dev.
`props.ini` hold the properties like endpoint URL and token etc. I also anticipate adding more properties to it in future
`logging.ini` emits the events to the console also to a file.

Todo: This application is written in python. I wanted to have the `event_transformer.py` and `send_to_udp.py` in its own package. somehow when I try to import the packages to `main.py` those packages were not recognized. I feel the way the project is set up ( as a subdirectory in the main project) is making it little challenging to do. I will puzzle this out in future Jira in this project.

pushyamig committed:
TLCALIPER-124 send caliper 1.1 events to Endpoint

pushyamig commented:
![](https://github.trello.services/images/mini-trello-icon.png) [TLCALIPER-124 Sending static events to an endpoint](https://trello.com/c/TZUKU5BA/1241-tlcaliper-124-sending-static-events-to-an-endpoint)

lsloan commented on src/simulate/event_transformer.py:
```diff
@@ -0,0 +1,25 @@
+
+    def event_transformer(self):
+        try:
+            json_event = json.loads(self.event)
```
I recommend using symbol names that more accurately represent their contents and purpose.  Instead of `json_event`, a better name is `eventData` or `eventObjects`.

pushyamig replied:
Done!

lsloan commented on src/simulate/event_transformer.py:
```diff
@@ -0,0 +1,25 @@
+            return None
+        data = json_event['data']
+        data[0]['eventTime'] = self.get_current_date_time_iso8601_format()
+        json_event['sendTime'] = self.get_current_date_time_iso8601_format()
```
Using better names...
`envelope = eventData`
`event = eventData['data'][0]`

Then the following assignments are more self-explanatory:
`envelope['sendTime'] = self.get_current_date_time_iso8601_format()`
`event['eventTime'] = self.get_current_date_time_iso8601_format()`

pushyamig replied:
Done!

lsloan reviewed (commented):
I'll review more soon.

pushyamig committed:
TLCALIPER-124 better naming in the Transformer class

pushyamig commented:
I figured out the problem with the python packages and import problem. Apparently, I see that python3 have different style of importing package vs python2
```
The project structure
simulate:
   pack_event_transformer:
        __init__.py
       Transformer.py
   main.py
```
```
main.py
from pack_event_transformer.Transformer import Transformer
.....
# some code
tranformer_instance = Transformer()
```

lsloan commented:
The import you described for Python 3 looks just like that I'm used to from Python 2.  What difference did you find?

Also, be careful with spelling the issue references.  You used "TLCAIPER-124" (without the "L") in the last commit.  If JIRA or GitHub is using those names to correlate issues and commits, it probably failed with that one.  Use copy and paste rather than retyping the name each time.

pushyamig commented:
I fixed the issue with the commit message.

I have a Python class `Transformer` in the module called `event_transformer.py`. I felt while importing the module this should work ` from event_transformer import Transformer`  but it threw an error saying no module by that name. Then I started looking up python3 specific content for package there I got a clue that it should be done something like this
`from event_transformer.Transformer import Transformer`.

I see the difference between py2 vs py3. For me, something that worked in Py2 did not work in the py3 so I have to look up and details were subtle 
one example, while sending the json events using requests module I changed in py3 as
`requests.post(url, json=data, headers=headers)`
vs py2 
`requests.post(url, data=data, headers=headers)`

lsloan commented:
Part of my surprise about the imports in Python 2 vs. Python 3 is the arrangement of `simulate/event_transformer/Transformer.py` vs. `simulate/event_transformer.py`. You didn't move the class into a directory in this PR.  You didn't do that until the next PR, #14.

The example you cited about differences with `requests.post()` is a change in the requests library, not a Python 2/3 issue.  (Maybe the latest version of requests for Python 2 doesn't include the `json` to `data` keyword argument change.)

lsloan commented on src/simulate/config/props.ini:
```diff
@@ -0,0 +1,3 @@
+[oauth]
```
Although `logging.config.fileConfig()` requires a `.ini` file, that format doesn't need to be used for all of this app's config files.  Consider using YAML for this file instead.

pushyamig replied:
Yes, I realized that yaml is better and changed it in the next PR #14

lsloan commented on src/simulate/config/logging.ini:
```diff
@@ -0,0 +1,27 @@
+[loggers]
```
Although I wrote earlier for `props.ini` below that logging config seems to require `.ini` files, that's not the case.  You can replace `logging.config.fileConfig()` with a combination of `yaml.safe_load()` and `logging.config.dictConfig()` to read the config from a YAML file.

lsloan commented on src/simulate/event_transformer.py:
```diff
@@ -0,0 +1,26 @@
+        self.event = event
+
+    def get_current_date_time_iso8601_format(self):
+        strftime = datetime.datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%f')
```
Rename your `strftime` variable to something else.  It's awkward for it to be named the same as the `strftime()` method.  Try `isoUtcNow` or `isoTimestamp` instead.

pushyamig commented:
Yes, while I was working in the PR, #14  I understood my problem with importing packages was and since I mentioned this problem in the PR, #13 I stated in that PR.

lsloan reviewed (commented):
I suggest a couple small changes, but it's otherwise OK.  Does it run well?  Do you see events in the endpoint as expected?

lsloan commented on src/simulate/README.md:
```diff
@@ -0,0 +1,15 @@
+ The Unizin Data Platform is intended to absorb the caliper events. UDP is development in progress and this application is taking caliper 1.1 events including envelope from flat files sends it to UDP
+
+### Running requirements
+1. The script is tested with Python 3.6.3
```
This is a silly-ish comment:  Something that annoys me is useful programs being referred to as "scripts".  To me, scripts are small, simple pieces of code that do a useful task, but don't have much logic to them.  Something like the program you're writing here is more complex than that and deserves to be called a "program" or if it's big enough (I think it is), "app".

pushyamig commented:
Yes, this sends the event to the endpoint http://lti.tools/caliper/event?key=udp_dev. I choose this over the openlrs-dev. 
The specific reason if you as I tried to send the events to the openlrs-dev but with the version of requests module I was using it thrown SSL certs error. I looked up what that means and I got to know that the openlrs-dev don't have the full stack of certs established and I need to add some certs while sending the request. I felt let's not go that route as UDP is the target and it might not have this specific problem.
In the past i have sent the caliper events to openlrs-dev with requests module ( version that supports py2) I did not have this problem. So a question that keeps popping up is something changed with modules with Py3.

lsloan commented:
We **_always_** have problems with HTTPS connections to our OpenLRS services.  It's because we use certificates signed by the U-M CA, not a well-known CA.  Whenever we need apps to communicate with our OpenLRS services, we always need to install a CA cert. on the app server and tell the library (cURL, etc.) to use that cert. for verification.  We did that with Lecture Capture and Problem Roulette.
